### DISCUSSION TITLE
Object implementation notes

### DISCUSSION THREAD

**Discussion by iainctduncan** [iainctduncan/scheme-for-max#109] [Ideas]
Capturing some stuff from the CMDist list to not forget while brainstorming up object implementation for the live coding toolkit

**Comment by iainctduncan** [comment]
From Bill:
 "stuff.scm has define-class and friends with some examples
in s7test.scm.  It has some similarities to CLOS."
"I wrote that about a decade ago, but as far as I can
remember, I just wrote it from scratch.  I used CLOS
(in cmn) back in the early 90's, so I had some idea
what I wanted."

**Reply by ulzu** [reply]
Do you maybe have some notes on how that set of functions (define-class, -generic, -method…) are to be glued together?
I'd love to try it out, it seems it might feel nice and tidy, but these macro definitions are too complex for me to unravel anything usable.

**Reply by iainctduncan** [reply]
Hi, I've actually been working on my own object system optimized for live/interactive  coding, so I didn't wind up going down this path. I will be publishing it over the next few months as part of my thesis. It's simpler and designed expressly to lower the number of calls one needs to make in working with objects as well as to simplify serializing the stuff you might want to save with a piece. I can share it with you it you like once it's a little further.

**Reply by ulzu** [reply]
I would be very interested!
I think I would benefit a lot from it for what I'm trying to build now (a little system for making ad-hoc multi-modal bindings to grid- and cc-controllers, I hope I'm up to the task!). I'm now passing around simple let-over-lambdas, but having a streamlined object system would be very convenient.
Plus, I'm sure it would be very instructive for anyone interested in s4m to see how you might be implementing it.

**Comment by iainctduncan** [comment]
From Rick:
FWIW The Common Music 2 code base is 100% CLOS.  The Grace code base uses Bil’s record structure to implement its Scheme pattern objects and it works quite well.

C Side...

pattern_tag = s7_make_c_type(s7, "<pattern>");

s7_define_function(s7, "make-pattern", ffi_make_pattern, 14, 0, false, "pattern constructor”);
[...]

Scheme Side...

(define-record pattern flags data length datum period value state 
  repeat returning counting traversing next mapr cache)


(define (%alloc-pattern)
  ;; flags data length datum period value state limit returning counting traversing next mapr cache
  (make-pattern 0 (list) #f +nad+ #f +nad+ +nad+ most-positive-fixnum #f #:periods #:depth-first
#f #f #f))

(define (initialize-pattern obj data for rep flags len dper getr mapr)
  (pattern-data-set! obj data)
  (pattern-length-set! obj len)
  (pattern-mapr-set! obj mapr)
  (pattern-next-set! obj getr)
  ;; map data to see if it is constant data or has subpatterns
  (let ((con? #t))
    (map-pattern-data (lambda (x) (if (pattern? x) (set! con? #f))) 
      obj)
    (if con? (set! flags (logior flags +constant-data+))))
   [...]
)
[...]
(define (make-cycle data . args)
  (unless (pair? data) (set! data (list data)))
  (with-optkeys (args for limit)
    (let ((obj (%alloc-pattern))
  (flags 0)
  (len (length data)))
      (initialize-pattern obj (cons data data) for limit
  flags len len next-in-cycle
  (lambda (fn obj) 
    (for-each fn (car (pattern-data obj)))))
      obj)))

[...]
